### ISSUE TITLE
unexpected stack underflow with u32split

### ISSUE THREAD

**Issue by grjte** [maticnetwork/miden#70]
## TL;DR: 

There seems to be an issue in one of the 2 following places:
-  a problem with how the stack depth is managed for operations that output more elements than they take as inputs
- a bug in the check_depth requirement in ```op_u32split``` + a lack of documentation about stack requirements for that op

## Description

According to the assembly ops specs [here](https://hackmd.io/YDbjUVHTRn64F4LPelC-NA?view#u32-operations), which specify a stack input for ```u32cast``` of [a, ...] and stack output of [b, ...] I expected the following minimal execution test to pass, but it fails with a ```stack underflow``` error.

```
#[test]
fn u32cast() {
    let script = compile("begin u32cast end");

    let inputs = build_inputs(&[1]);
    let trace = execute(&script, &inputs).unwrap();
    let last_state = trace.last_stack_state();
    let expected_state = build_stack_state(&[1]);

    assert_eq!(expected_state, last_state);
}
```

Essentially what's happening is that when the processor does a [```u32split```](https://github.com/maticnetwork/miden/blob/next/processor/src/operations/u32_ops.rs) & attempts to set ```stack[1]``` with the low 32-bit values, it panics, because the current (and required) stack depth is only 1. If I add an extra input to the stack then everything works fine, and that input isn't touched.

Relevant backtrace:
```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: miden_processor::stack::Stack::set
             at ./src/stack/mod.rs:99:9
   3: miden_processor::operations::u32_ops::<impl miden_processor::Process>::op_u32split
             at ./src/operations/u32_ops.rs:19:9
```

## Question

This seems to be a bug either in the way the stack depth is managed for operations that output more elements than they take as inputs or a bug in the ```check_depth``` assertion in ```op_u32split```. What gave me pause was that in the [```u32split``` unit test in the processor](https://github.com/maticnetwork/miden/blob/next/processor/src/operations/u32_ops.rs) you're initializing the stack with 2 values, which suggests that maybe the current behaviour is intentional but wasn't documented yet (or is in a place I haven't seen).

What is supposed to be happening here?

**Comment by bobbinth** [comment]
Great catch! The bug is in the way stack depth is managed. Specifically, we increment the stack depth when we call `stack.shift_right()` but the `stack.set()` checks for stack depth before it can set a given item. So, for operations which shift the stack to the right, if `stack.set()` comes before `stack.shift_right()`, we'll get a panic.

One way to address this is for operations which add an item to the stack is to make sure we always call `stack.shift_right()` before we call `stack.set()`. But there might be other (better?) ways of solving this.

Regarding unit test, the reason I initialize the stack with two values to make sure that `u32split` doesn't overwrite the value it's not supposed to touch. But it hides another bug - so, I kind of outsmarted myself here :)

**Comment by grjte** [comment]
> the reason I initialize the stack with two values was to make sure that u32split doesn't overwrite the value it's not supposed to touch. But it hides another bug - so, I kind of outsmarted myself here :)

In fairness, that one's important too!

> One way to address this for operations which add an item to the stack is to make sure we always call stack.shift_right() before we call stack.set()

Hm, yeah, I don't like these sorts of "order matters" patterns, unless they're absolutely necessary like Checks-Effects-Interactions in Solidity. However, I don't have a better solution at the moment, and this would resolve the issue so things can move forwards. At least it's a controlled & limited setting, so it's ok with solid testing. Worth letting it marinate to see if we come up with something better in the future though

**Comment by grjte** [comment]
For now, I'll fix this with that proposed solution as I continue with the u32ops tests, if that's ok?

**Comment by bobbinth** [comment]
Yes - I think this works! I can't think of anything better at the moment either.

**Comment by bobbinth** [comment]
fixed in #75
